File: python/googledatastore/datastore_emulator.py
#
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""Python wrapper for the Cloud Datastore emulator."""

__author__ = 'eddavisson@google.com (Ed Davisson)'


import httplib
import logging
import os
import shutil
import socket
import subprocess
import tempfile
import time
import zipfile

from googledatastore import connection
import httplib2
import portpicker


_DEFAULT_EMULATOR_OPTIONS = ['--testing']


class DatastoreEmulatorFactory(object):
  """A factory for constructing DatastoreEmulator objects."""

  def __init__(self, working_directory, emulator_zip, java=None):
    """Constructs a factory for building datastore emulator instances.

    Args:
      working_directory: path to a directory where temporary files will be
          stored
      emulator_zip: path to the emulator zip file
      java: path to a java executable
    """
    self._working_directory = working_directory

    self._emulators = {}

    # Extract the emulator.
    zipped_file = zipfile.ZipFile(emulator_zip)
    if not os.path.isdir(self._working_directory):
      os.mkdir(self._working_directory)
    zipped_file.extractall(self._working_directory)

    self._emulator_dir = os.path.join(self._working_directory,
                                      'cloud-datastore-emulator')
    self._emulator_cmd = os.path.join(self._emulator_dir,
                                      'cloud_datastore_emulator')
    os.chmod(self._emulator_cmd, 0700)  # executable

    # Make the emulator use our copy of Java.
    if java:
      os.environ['JAVA'] = java

  def Get(self, project_id):
    """Returns an existing emulator instance for the provided project_id.

    If an emulator instance doesn't yet exist, it creates one.

    Args:
      project_id: project ID

    Returns:
      a DatastoreEmulator
    """
    if project_id in self._emulators:
      return self._emulators[project_id]

    emulator = self.Create(project_id)
    self._emulators[project_id] = emulator
    return emulator

  def Create(self, project_id, start_options=None, deadline=10):
    """Creates an emulator instance.

    This method will wait for up to 'deadline' seconds for the emulator to
    start.

    Args:
      project_id: project ID
      start_options: a list of additional command-line options to pass to the
          emulator 'start' command
      deadline: number of seconds to wait for the datastore to respond

    Returns:
      a DatastoreEmulator

    Raises:
      IOError: if the emulator could not be started within the deadline
    """
    return DatastoreEmulator(self._emulator_cmd, self._working_directory,
                             project_id, deadline, start_options)

  def __del__(self):
    # Delete temp files.
    shutil.rmtree(self._emulator_dir)


class DatastoreEmulator(object):
  """A Datastore emulator."""

  def __init__(self, emulator_cmd, working_directory, project_id, deadline,
               start_options):
    """Constructs a DatastoreEmulator.

    Clients should use DatastoreEmulatorFactory to construct DatastoreEmulator
    instances.

    Args:
      emulator_cmd: path to cloud_datastore_emulator
      working_directory: directory file where temporary files will be stored
      project_id: project ID
      deadline: number of seconds to wait for the datastore to start
      start_options: a list of additional command-line options to pass to the
          emulator 'start' command

    Raises:
      IOError: if the emulator failed to start within the deadline
    """
    self._project_id = project_id
    self._emulator_cmd = emulator_cmd
    self._http = httplib2.Http()
    self.__running = False

    self._tmp_dir = tempfile.mkdtemp(dir=working_directory)
    self._project_directory = os.path.join(self._tmp_dir, self._project_id)
    p = subprocess.Popen([emulator_cmd,
                          'create',
                          '--project_id=%s' % self._project_id,
                          self._project_directory])
    if p.wait() != 0:
      raise IOError('could not create project in directory: %s'
                    % self._project_directory)

    # Start the emulator and wait for it to start responding to requests.
    port = portpicker.PickUnusedPort()
    self._host = 'http://localhost:%d' % port
    cmd = [self._emulator_cmd, 'start', '--port=%d' % port]
    cmd.extend(_DEFAULT_EMULATOR_OPTIONS)
    if start_options:
      cmd.extend(start_options)
    cmd.append(self._project_directory)
    subprocess.Popen(cmd)
    if not self._WaitForStartup(deadline):
      raise IOError('emulator did not respond within %ds' % deadline)
    endpoint = '%s/v1/projects/%s' % (self._host, self._project_id)
    self.__datastore = connection.Datastore(project_endpoint=endpoint)
    self.__running = True

  def GetDatastore(self):
    """Returns a googledatatsore.Datastore that is connected to the emulator."""
    return self.__datastore

  def _WaitForStartup(self, deadline):
    """Waits for the emulator to start.

    Args:
      deadline: deadline in seconds

    Returns:
      True if the emulator responds within the deadline, False otherwise.
    """
    start = time.time()
    sleep = 0.05

    def Elapsed():
      return time.time() - start

    while True:
      try:
        response, _ = self._http.request(self._host)
        if response.status == 200:
          logging.info('emulator responded after %f seconds', Elapsed())
          return True
      except (socket.error, httplib.ResponseNotReady):
        pass
      if Elapsed() >= deadline:
        # Out of time; give up.
        return False
      else:
        time.sleep(sleep)
        sleep *= 2

  def Clear(self):
    """Clears all data from the emulator instance.

    Returns:
      True if the data was successfully cleared, False otherwise.
    """
    headers = {'Content-length': '0'}
    response, _ = self._http.request('%s/reset' % self._host, method='POST',
                                     headers=headers)
    if response.status == 200:
      return True
    else:
      logging.warning('failed to clear emulator; response was: %s', response)

  def Stop(self):
    """Stops the emulator instance."""
    if not self.__running:
      return
    logging.info('shutting down the emulator running at %s', self._host)
    headers = {'Content-length': '0'}
    response, _ = self._http.request('%s/shutdown' % self._host,
                                     method='POST', headers=headers)
    if response.status != 200:
      logging.warning('failed to shut down emulator; response: %s', response)

    self.__running = False
    # Delete temp files.
    shutil.rmtree(self._tmp_dir)

  def __del__(self):
    # If the user forgets to call Stop()
    logging.warning('emulator shutting down due to '
                    'DatastoreEmulator object deletion')
    self.Stop()

File: java/datastore/src/main/java/com/google/datastore/v1/client/DatastoreEmulator.java
/*
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.datastore.v1.client;

import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.HttpRequestFactory;
import com.google.api.client.http.HttpResponse;
import com.google.api.client.http.UrlEncodedContent;
import com.google.api.client.util.Preconditions;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * An extension to {@link Datastore} that provides lifecycle management for a development datastore
 * server.
 *
 * In order to use a development datastore for a JUnit 4 test you might do something like this:
 *
 * <pre>
 * public class MyTest {
 *
 *   static LocalDevelopmentDatastore datastore;
 *
 *   {@literal @}BeforeClass
 *   public static void startLocalDatastore() throws LocalDevelopmentDatastoreException {
 *     DatastoreOptions opts = new DatastoreOptions.Builder()
 *         .localHost("localhost:8080")
 *         .projectId("my-project-id")
 *         .build();
 *     datastore = LocalDevelopmentDatastoreFactory.get().create(opts);
 *     datastore.start("/usr/local/gcd-sdk", "my-project-id");
 *   }
 *
 *   {@literal @}Before
 *   public void setUp() throws LocalDevelopmentDatastoreException {
 *     datastore.clear();
 *   }
 *
 *   {@literal @}AfterClass
 *   public static void stopLocalDatastore() throws LocalDevelopmentDatastoreException {
 *     datastore.stop();
 *   }
 *
 *   {@literal @}Test
 *   public void testFoo1() { }

 *   {@literal @}Test
 *   public void testFoo2() { }
 *
 * }
 * </pre>
 */
public class LocalDevelopmentDatastore extends Datastore {
  private static final int STARTUP_TIMEOUT_SECS = 30;

  private final String host;
  private LocalDevelopmentDatastoreOptions localDevelopmentOptions;

  /** Internal state lifecycle management. */
  enum State {NEW, STARTED, STOPPED}

  private volatile State state = State.NEW;

  private File projectDirectory;

  LocalDevelopmentDatastore(RemoteRpc rpc, String localHost,
      LocalDevelopmentDatastoreOptions localDevelopmentOptions) {
    super(rpc);
    this.host = "http://" + localHost;
    this.localDevelopmentOptions = localDevelopmentOptions;
  }

  /**
   * Clears all data in the Datastore.
   *
   * @throws LocalDevelopmentDatastoreException
   */
  public void clear() throws LocalDevelopmentDatastoreException {
    HttpRequestFactory client = remoteRpc.getHttpRequestFactory();
    try {
      Map<String, String> params = new HashMap<String, String>();
      params.put("action", "Clear Datastore");
      UrlEncodedContent content = new UrlEncodedContent(params);
      GenericUrl url = new GenericUrl(host + "/_ah/admin/datastore");
      HttpResponse httpResponse = client.buildPostRequest(url, content).execute();
      if (!httpResponse.isSuccessStatusCode()) {
        throw new LocalDevelopmentDatastoreException(
            "Clear Datastore returned http status " + httpResponse.getStatusCode());
      }
    } catch (IOException e) {
      throw new LocalDevelopmentDatastoreException(
          "Exception trying to clear the dev datastore", e);
    }
  }

  /**
   * Starts the local datastore. It is the caller's responsibility to call {@link #stop}. Note that
   * receiving an exception does not indicate that the server did not start. We recommend calling
   * {@link #stop} to ensure the server is not running regardless of the result of this method.
   *
   * @param sdkPath The path to the GCD SDK, eg /usr/local/dev/gcd
   * @param projectId The GCD project ID
   * @param cmdLineOptions Command line options to pass to the script that launches the dev server
   * @throws LocalDevelopmentDatastoreException If {@link #start} has already been called or the
   * server does not start successfully.
   */
  public synchronized void start(String sdkPath, String projectId, String... cmdLineOptions)
      throws LocalDevelopmentDatastoreException {
    Preconditions.checkNotNull(sdkPath, "sdkPath cannot be null");
    Preconditions.checkNotNull(projectId, "projectId cannot be null");
    Preconditions.checkState(state == State.NEW, "Cannot call start() more than once.");
    try {
      startDatastoreInternal(sdkPath, projectId, cmdLineOptions);
      state = State.STARTED;
    } finally {
      if (state != State.STARTED) {
        // If we're not able to start the server we don't want people trying again. Just move it
        // straight to the STOPPED state.
        state = State.STOPPED;
      }
    }
  }

  void startDatastoreInternal(String sdkPath, String projectId, String... cmdLineOptions)
      throws LocalDevelopmentDatastoreException {
    File projectDirectory = createProjectDirectory(sdkPath, projectId);
    List<String> cmd = new ArrayList<String>(
        Arrays.asList("./gcd.sh", "start", "--allow_remote_shutdown", "--store_on_disk=false"));
    cmd.addAll(Arrays.asList(cmdLineOptions));
    cmd.add(projectDirectory.getPath());
    final Process gcdStartProcess;
    try {
      gcdStartProcess = newGcdProcess(sdkPath, cmd).start();
    } catch (IOException e) {
      throw new LocalDevelopmentDatastoreException("Could not start dev server", e);
    }
    // Ensure we don't leak the stub instance if tests end prematurely.
    Runtime.getRuntime().addShutdownHook(new Thread() {
      @Override
      public void run() {
        gcdStartProcess.destroy();
    }
    });
    StartupMonitor monitor = new StartupMonitor(gcdStartProcess.getInputStream());
    try {
      monitor.start();
      if (!monitor.startupCompleteLatch.await(STARTUP_TIMEOUT_SECS, TimeUnit.SECONDS)) {
        throw new LocalDevelopmentDatastoreException("Dev server did not start within 30 seconds");
      }
      if (!monitor.success) {
        throw new LocalDevelopmentDatastoreException("Server did not start normally");
      }
    } catch (InterruptedException e) {
      // not sure why this would happen
      throw new LocalDevelopmentDatastoreException("Received an interrupt", e);
    }
  }

  private File createProjectDirectory(String sdkPath, String projectId)
      throws LocalDevelopmentDatastoreException {
    try {
      projectDirectory = Files.createTempDirectory("local-development-datastore").toFile();
    } catch (IOException e) {
      throw new LocalDevelopmentDatastoreException("Could not create project tmp directory", e);
    }
    List<String> cmd = Arrays.asList("./gcd.sh", "create", "--project_id=" + projectId,
        projectDirectory.getPath());
    try {
      int retCode = newGcdProcess(sdkPath, cmd).start().waitFor();
      if (retCode != 0) {
        throw new LocalDevelopmentDatastoreException(
            String.format("Could not create project (retcode=%d)", retCode));
      }
    } catch (IOException e) {
      throw new LocalDevelopmentDatastoreException("Could not create project", e);
    } catch (InterruptedException e) {
      throw new LocalDevelopmentDatastoreException("Received an interrupt", e);
    }
    return projectDirectory;
  }

  private ProcessBuilder newGcdProcess(String sdkPath, List<String> cmd) {
    ProcessBuilder builder = new ProcessBuilder(cmd);
    builder.directory(new File(sdkPath));
    builder.redirectErrorStream(true);
    builder.environment().putAll(localDevelopmentOptions.getEnvVars());
    return builder;
  }

  /**
   * Stops the local datastore. Multiple calls are allowed.
   *
   * @throws LocalDevelopmentDatastoreException
   */
  public synchronized void stop() throws LocalDevelopmentDatastoreException {
    // We intentionally don't check the internal state. If people want to try and stop the server
    // multiple times that's fine.
    stopDatastoreInternal();
    if (state != State.STOPPED) {
      state = State.STOPPED;
      if (projectDirectory != null) {
        try {
          Process process =
              new ProcessBuilder("rm", "-r", projectDirectory.getAbsolutePath()).start();
          if (process.waitFor() != 0) {
            throw new IOException("Temp directory delete exited with " + process.exitValue());
          }
        } catch (IOException | InterruptedException e) {
          throw new IllegalStateException("Project directory wipe failed.", e);
        }
      }
    }
  }

  void stopDatastoreInternal() throws LocalDevelopmentDatastoreException {
    // No need to kill the process we started, this function will take care of it.
    HttpRequestFactory client = remoteRpc.getHttpRequestFactory();
    Map<String, String> params = new HashMap<String, String>();
    UrlEncodedContent content = new UrlEncodedContent(params);
    GenericUrl url = new GenericUrl(host + "/_ah/admin/quit");
    try {
      HttpResponse httpResponse = client.buildPostRequest(url, content).execute();
      if (!httpResponse.isSuccessStatusCode()) {
        throw new LocalDevelopmentDatastoreException(
            "Request to shutdown local datastore returned http error code "
            + httpResponse.getStatusCode());
      }
    } catch (IOException e) {
      throw new LocalDevelopmentDatastoreException(
          "Exception trying to stop the dev datastore", e);
    }
  }

  /**
   * Monitors the provided input stream for evidence that the dev server has started successfully
   * and redirects the output of the dev server process to sysout for this process.
   */
  static class StartupMonitor extends Thread {

    private final InputStream inputStream;
    private volatile boolean success = false;
    /** This latch will reach 0 once server startup has completed. */
    private final CountDownLatch startupCompleteLatch = new CountDownLatch(1);

    StartupMonitor(InputStream inputStream) {
      this.inputStream = inputStream;
      setDaemon(true);
    }

    @Override
    public void run() {
      try {
        BufferedReader br = new BufferedReader(new InputStreamReader(inputStream));
        String line;
        while ((line = br.readLine()) != null) {
          // redirect to sysout for our process
          System.out.println(line);
          if (!success && line.contains("Dev App Server is now running")) {
            success = true;
            startupCompleteLatch.countDown();
          }
        }
      } catch (IOException ioe) {
        if (!success) {
          System.err.println("Received an IOException before the dev server startup completed. "
              + "Dev server is in an unknown state.");
        } else {
          // We got an exception after the server started successfully. We'll lose the ability
          // to log the output of the dev server but there's no need to shut anything down.
          System.err.println("Received an exception handling output from the dev server. "
              + "Logging will stop but the dev server is probably ok.");
        }
        ioe.printStackTrace();
      } finally {
        if (!success) {
          // Either the stream is closed (indicates server shut down) or we received an Exception
          // while processing the stream contents. Either way we can tell the calling thread to stop
          // waiting.
          startupCompleteLatch.countDown();
        }
      }
    }
  }
}
